notebooks/utils.py

```diff
@@ -0,0 +1,127 @@
+        
+        self.simple_funcs = {
+            'exponential':torch.exp,
+            'sinusoidal':torch.sin
+        }
+        
+        self.debug = False
+        
+        self.params = {
+            'domain_min': -1,
+            'domain_max': 3,
+            'num_points': 81,
+            'function': 'exponential',
+            'hidden': 1,
+            'width': 5,
+            'nepoch': 100,
+            'lr': 0.002,
+            'momentum': 0.9,
+            'swarm_size': 20
+            
+        }
+        
+        self.simulations = []
+        
+        
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        
+        
+    def reset_swarm(self):
+        self.bees = [{'loss_list':[], 'pred_list': []} for i in range(self.params['swarm_size'])]
+        
+    def rebuild_predictor(self):
+        self.xt = torch.linspace(self.params['domain_min'], self.params['domain_max'], self.params['num_points']).unsqueeze(-1)
+        self.chosen_func = self.simple_funcs[self.params['function']]
+        self.yt = self.chosen_func(self.xt)
+        self.xd = self.xt.detach()
+        self.yd = self.yt.detach()
+        self.get_pred = self.make_predictor(self.xt, self.yt)
+        self.reset_swarm()
+        
+
+    def make_net(self, hidden_depth, width):
+        assert hidden_depth >= 1
+        yield nn.Linear(1, width)
+        yield nn.ReLU()
+        for i in range(hidden_depth-1):
+            yield nn.Linear(width, width)
+            yield nn.ReLU()
+        yield nn.Linear(width, 1)
+    
+    def make_predictor(self, x, y):
+        xt = x.clone()
+        yt = y.clone()
+        def get_pred(num, hidden = self.params['hidden'], width = self.params['width'], nepoch=self.params['nepoch'], lr=self.params['lr'], momentum=self.params['momentum'], debug=self.debug):
+            net = nn.Sequential(*self.make_net(hidden, width))
+            lossfunc = nn.MSELoss()
+            optimizer = optim.SGD(net.parameters(), lr=lr, momentum=momentum)
+
+            og_loss = lossfunc(net(xt), yt)
+            for epoch in range(nepoch):
+                optimizer.zero_grad()
+                ypred = net(xt)
+                self.bees[num]['pred_list'].append(np.array(ypred.detach()))
+                loss = lossfunc(ypred, yt)
+                self.bees[num]['loss_list'].append(loss.item())
+                if debug: print(epoch, loss)
+                loss.backward()
+                optimizer.step()
+                
+            if debug: print(f"First loss {og_loss} v final {loss}")
```

Comment:
here's a good example - you've defined self.debug before but you've taken debug in here too.

Reply:
Yeah, I should really get my head around that. I haven't used debggin much before. in my code.